src/storage/mvcc/reader.rs

```diff
@@ -0,0 +1,187 @@
+        self.seek_write_impl(key, ts, true)
+    }
+
+    fn seek_write_impl(&mut self, key: &Key, ts: u64, reverse: bool) -> Result<Option<(u64, u64)>> {
+        let (k, v) = {
+            if self.write_cursor.is_none() {
+                self.write_cursor = Some(try!(self.snapshot.iter_cf(CF_WRITE)));
+            }
+            let mut cursor = self.write_cursor.as_mut().unwrap();
+            let ok = if reverse {
+                try!(cursor.reverse_seek(&key.append_ts(ts)))
+            } else {
+                try!(cursor.seek(&key.append_ts(ts)))
+            };
+            if !ok {
+                return Ok(None);
+            }
+            (cursor.key().to_vec(), cursor.value().to_vec())
+        };
+        let (k, s, c) = try!(self.decode_write(k, v));
+        if &k == key {
+            Ok(Some((s, c)))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn decode_write(&self, key: Vec<u8>, value: Vec<u8>) -> Result<(Key, u64, u64)> {
+        let k = Key::from_encoded(key);
+        let commit_ts = try!(k.decode_ts());
+        let key = try!(k.truncate_ts());
+        let start_ts = try!(value.as_slice().decode_var_u64());
+        Ok((key, start_ts, commit_ts))
+    }
+
+    pub fn get(&mut self, key: &Key, ts: u64) -> Result<Option<Value>> {
+        // Check for locks that signal concurrent writes.
+        if let Some(lock) = try!(self.load_lock(key)) {
+            if lock.ts <= ts {
+                // There is a pending lock. Client should wait or clean it.
+                return Err(Error::KeyIsLocked {
+                    key: try!(key.raw()),
+                    primary: lock.primary,
+                    ts: lock.ts,
+                });
+            }
+        }
+        match try!(self.seek_write(key, ts)) {
+            Some((start_ts, _)) => self.load_data(key, start_ts),
+            None => Ok(None),
+        }
+    }
+
+    pub fn get_txn_commit_ts(&mut self, key: &Key, start_ts: u64) -> Result<Option<u64>> {
+        match try!(self.reverse_seek_write(key, start_ts)) {
+            Some((s, commit_ts)) if s == start_ts => Ok(Some(commit_ts)),
+            _ => Ok(None),
+        }
+    }
+
+    fn create_data_cursor(&mut self) -> Result<()> {
+        if self.data_cursor.is_none() {
+            self.data_cursor = Some(try!(self.snapshot.iter()));
+        }
+        Ok(())
+    }
+
+    pub fn seek(&mut self, mut key: Key, ts: u64) -> Result<Option<(Key, Value)>> {
+        try!(self.create_data_cursor());
+
+        loop {
+            key = {
+                let mut cursor = self.data_cursor.as_mut().unwrap();
+                if !try!(cursor.seek(&key)) {
+                    return Ok(None);
+                }
+                try!(Key::from_encoded(cursor.key().to_vec()).truncate_ts())
+            };
+            if let Some(v) = try!(self.get(&key, ts)) {
+                return Ok(Some((key, v)));
+            }
+            key = key.append_ts(0);
```

Comment:
so why reverse_seek seems return previous version of the current key, not previous key?